# Model Evaluation Exercises

In [1]:
import numpy as np
import pandas as pd
from model_evaluation_functions import evaluation_metrics

## 1. Artisanal Evaluation
Given the following confusion matrix, evaluate (_by hand_) the model's performance.

|               | actual cat | actual dog |
|:------------  |-----------:|-----------:|
| predicted cat |         34 |          7 |
| predicted dog |         13 |         46 |

In [2]:
from IPython.display import Image
print("Artisanal Evaluation")
# Image(filename='./evaluations-by-hand.png', width=400, height=400)

Artisanal Evaluation


Context:

    On earth 617, the only animals are cats and dogs. Cats and dogs roam freely on the earth. Certain human communities can only be around dogs, because they're allergic to cats. Cats must be kept away from these communities at all costs, or they will sneeze louder than 12 Saturn V Rockets. 

    They've recruited Chris and the gang from good ol' earth 616 to help them evaluate their defense model. The model predicts whether an animal is a cat or not a cat. If the model predicts a cat, industrial-sized presentation lasers will point at the ground and lead the cats away from the community. Dogs ignore the lasers and mark their territory. If a cat enters the community, well, RIP ear drums.

In the context of this problem, what is a false positive (Type I Error)?
> In the context of is this problem, a False Positive means that the model __predicted a picture to be a Cat__, but it was __actually a Dog__.
- _Close call_

In the context of this problem, what is a false negative (Type II Error)?
> In the context of is this problem, a False Negative means that the model __predicted a picture to be a Dog__, but it was __actually a Cat__.
- _RIP eardrums_

How would you describe this model?
> This model predicts whether an animal is a cat or not a cat.
- "Not a cat" == Dog

## 2. C3 Rubber Duck Manufacturer
You are working as a data scientist working for Codeup Cody Creator (C3 for short), a rubber-duck manufacturing plant.

Unfortunately, some of the rubber ducks that are produced will have defects. Your team has built several models that try to predict those defects.
Use the predictions dataset and pandas to help answer the following questions:

An internal team wants to investigate the cause of the manufacturing defects. They tell you that they want to identify as many of the ducks that have a defect as possible.

### Acquire the data

In [3]:
# Load c3's baseline model data - 3 models
df_c3 = pd.read_csv('./c3.csv')

In [4]:
# Create confusion matrices for each model to evaluate their performance.
model1_outcome_matrix = pd.crosstab(df_c3.model1, df_c3.actual)  # model 1
model2_outcome_matrix = pd.crosstab(df_c3.model2, df_c3.actual)  # model 2
model3_outcome_matrix = pd.crosstab(df_c3.model3, df_c3.actual)  # model 3

### Evaluate each model's performance

#### Model 1 Evaluation

In [5]:
model1_outcome_matrix

actual     Defect  No Defect
model1                      
Defect          8          2
No Defect       8        182

In [6]:
evaluation_metrics(model1_outcome_matrix)

Model Evaluation
----------------
Accuracy               95.00%
Recall                 80.00%
Precision              50.00%
Specificity            95.79%


#### Model 2 Evaluation

In [7]:
model2_outcome_matrix

actual     Defect  No Defect
model2                      
Defect          9         81
No Defect       7        103

In [8]:
evaluation_metrics(model2_outcome_matrix)

Model Evaluation
----------------
Accuracy               56.00%
Recall                 10.00%
Precision              56.25%
Specificity            93.64%


#### Model 3 Evaluation

In [9]:
model3_outcome_matrix

actual     Defect  No Defect
model3                      
Defect         13         86
No Defect       3         98

In [10]:
evaluation_metrics(model3_outcome_matrix)

Model Evaluation
----------------
Accuracy               55.50%
Recall                 13.13%
Precision              81.25%
Specificity            97.03%


1. Which evaluation metric would be appropriate here?

>__Answer__: The most appropriate metric for C3's business problem is **Recall**.
The model identifies defective or non-defective rubber ducks. If the model predicts "no defect" when there is a "defect" (Type II Error), our customer receives a defective rubber ducky.

>__Reasoning__: Evaluating ducks classified as "no-defect" (True Positives and False Negatives) helps us evaluate the model's performance.
- If the recall rate is high, the model can determine what a __True__ non-defective rubber duck is.
- If the recall rate is low, the model is sending out truckloads of defective rubber ducks.


2. Which model would be the best fit for this use case?
> Model 1 would be the best fit for this use case because it has the __highest__ _recall_.

Model 1 Evaluation
- Accuracy               95.00%
- Recall                 80.00%
- Precision              50.00%
- Specificity            95.79%

   
        Recently several stories in the local news have come out highlighting customers who received a rubber duck with a defect, and portraying C3 in a bad light. The PR team has decided to launch a program that gives customers with a defective duck a vacation to Hawaii.
> DANG. Talk about upholding their reputation. They've put the pressure on... they've hired one of the best in the business, `IGOT this`. I need to minimize those expensive vacations.

        They need you to predict which ducks will have defects, but tell you they really don't want to accidentally give out a vacation package when the duck really doesn't have a defect. Which evaluation metric would be appropriate here?
> __Answer__: *__Precision__* is the appropriate metric to evaluate the models.

>__Reasoning__: Evaluating ducks classified as "defect" (True Positives and False Positives) helps us evaluate the model's performance. Due to C3's PR stunt, we also need to make sure that people don't get a defective free duck AND get a vacation to Hawaii.
- Note: IRL customers would need to verify their claim. If C3 uses computer vision, its manufacturing plant would have a frame and timestamp when the customers' duck was evaluated.

    Which model would be the best fit for this use case?
> Model 3 would be the best fit for this use case because it has the __highest__ *precision*.

Model 3 Evaluation
- Accuracy               55.50%
- Recall                 81.25%
- Precision              81.25%
- Specificity            97.03%

## 3. Gives You Paws
You are working as a data scientist for Gives You Paws ™, a subscription based service that shows you cute pictures of dogs or cats (or both for an additional fee).

At Gives You Paws, anyone can upload pictures of their cats or dogs. The photos are then put through a two step process. First an automated algorithm tags pictures as either a cat or a dog (Phase I). Next, the photos that have been initially identified are put through another round of review, possibly with some human oversight, before being presented to the users (Phase II).

Several models have already been developed with the data, and you can find their results here.

Given this dataset, use pandas to create a baseline model (i.e. a model that just predicts the most common class) and answer the following questions:

In [11]:
df_paws = pd.read_csv('./gives_you_paws.csv')

In [12]:
df_paws.head()

actual model1 model2 model3 model4
0    cat    cat    dog    cat    dog
1    dog    dog    cat    cat    dog
2    dog    cat    cat    cat    dog
3    dog    dog    dog    cat    dog
4    cat    cat    cat    dog    dog

In [17]:
df_paws['baseline'] = df_paws.actual.value_counts().index[0]

In [24]:
df_paws.baseline.value_counts()

dog    5000
Name: baseline, dtype: int64

In [18]:
baseline_matrix = pd.crosstab(df_paws.baseline, df_paws.actual) # baseline
model1_matrix = pd.crosstab(df_paws.model1, df_paws.actual)  # model 1
model2_matrix = pd.crosstab(df_paws.model2, df_paws.actual)  # model 2
model3_matrix = pd.crosstab(df_paws.model3, df_paws.actual)  # model 3
model4_matrix = pd.crosstab(df_paws.model4, df_paws.actual)  # model 4

#### Baseline Evaluation

In [34]:
tp, tn = baseline_matrix.values.ravel()

In [37]:
precision = round(tn / (tp + tn), 2)
print(f"Accuracy {precision:.2%}")

Accuracy 65.00%


#### Model 1 Evaluation

In [38]:
model1_matrix

actual   cat   dog
model1            
cat     1423   640
dog      323  2614

In [39]:
evaluation_metrics(model1_matrix)

Model Evaluation
----------------
Accuracy               80.74%
Recall                 68.98%
Precision              81.50%
Specificity            89.00%


#### Model 2 Evaluation

In [40]:
model2_matrix

actual   cat   dog
model2            
cat     1555  1657
dog      191  1597

In [41]:
evaluation_metrics(model2_matrix)

Model Evaluation
----------------
Accuracy               63.04%
Recall                 48.41%
Precision              89.06%
Specificity            89.32%


#### Model 3 Evaluation

In [42]:
model3_matrix

actual  cat   dog
model3           
cat     893  1599
dog     853  1655

In [43]:
evaluation_metrics(model3_matrix)

Model Evaluation
----------------
Accuracy               50.96%
Recall                 35.83%
Precision              51.15%
Specificity            65.99%


#### Model 4 Evaluation

In [44]:
model4_matrix

actual   cat   dog
model4            
cat      603   144
dog     1143  3110

In [45]:
evaluation_metrics(model4_matrix)

Model Evaluation
----------------
Accuracy               74.26%
Recall                 80.72%
Precision              34.54%
Specificity            73.12%


    In terms of accuracy, how do the various models compare to the baseline model? Are any of the models better than the baseline?
> Model 1 and Model 4 outperform the baseline model. Model 2 and Model 3 preform worse than the baseline model.

    Suppose you are working on a team that solely deals with dog pictures.
    1. Which of these models would you recommend for Phase I?
    2. For Phase II?
    
    
    Suppose you are working on a team that solely deals with cat pictures.
    1. Which of these models would you recommend for Phase I?
    2. For Phase II?

## 4.
Follow the links below to read the documentation about each function, then apply those functions to the data from the previous problem.

    sklearn.metrics.accuracy_score
    sklearn.metrics.precision_score
    sklearn.metrics.recall_score
    sklearn.metrics.classification_report